frame/support/src/storage/mod.rs

```diff
@@ -733,6 +791,61 @@ impl<T> Iterator for PrefixIterator<T> {
 	}
 }
 
+/// Iterate over a prefix and decode raw_key into `T`.
+///
+/// If any decoding fails it skips it and continues to the next key.
+pub struct KeyPrefixIterator<T> {
+	prefix: Vec<u8>,
+	previous_key: Vec<u8>,
+	/// If true then value are removed while iterating
+	drain: bool,
+	/// Function that take `raw_key_without_prefix` and decode `T`.
+	/// `raw_key_without_prefix` is the raw storage key without the prefix iterated on.
+	closure: fn(&[u8]) -> Result<T, codec::Error>,
+}
+
+impl<T> KeyPrefixIterator<T> {
+	/// Mutate this iterator into a draining iterator; items iterated are removed from storage.
+	pub fn drain(mut self) -> Self {
+		self.drain = true;
+		self
+	}
+}
+
+impl<T> Iterator for KeyPrefixIterator<T> {
+	type Item = T;
+
+	fn next(&mut self) -> Option<Self::Item> {
+		loop {
+			let maybe_next = sp_io::storage::next_key(&self.previous_key)
+				.filter(|n| n.starts_with(&self.prefix));
+			break match maybe_next {
```

Comment:
Huh, I'd thought that `break` only accepted labels, not values. TIL. Could be worth adding a comment explaining what's happening there.

Likewise, could be worth mentioning that this is a `match` not a `map` so that `continue` works properly within.

Reply:
I actually copied code from the `Iterator` implementation of `PrefixIterator`. In all honesty, I wouldn't have written code like this, and I would have at least turned this `break` into a `return`. But yeah I'll try to think of ways to simplify this.